In [2]:
import pandas as pd
import numpy as np

In [3]:
exploratory = pd.read_csv("/Users/gohjunwei/Dropbox/4 - upworthy (mike)/Upworthy data documentation/upworthy-archive-datasets/upworthy-archive-confirmatory-packages-03.12.2020.csv")

/var/folders/1j/m3_4482d21x5c4l8wtby4flr0000gn/T/ipykernel_42406/3625148448.py:1: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  exploratory = pd.read_csv("/Users/gohjunwei/Dropbox/4 - upworthy (mike)/Upworthy data documentation/upworthy-archive-datasets/upworthy-archive-confirmatory-packages-03.12.2020.csv")


We begin by looking at the first few rows of the dataset to get a sense of what the data looks like.

In [4]:
exploratory

Unnamed: 0               created_at               updated_at  \
0               11  2014-11-20 11:33:26.475  2016-04-02 16:25:54.046   
1               12  2014-11-20 15:00:01.032  2016-04-02 16:25:54.128   
2               13  2014-11-20 11:33:51.973  2016-04-02 16:25:54.069   
3               14  2014-11-20 11:34:12.107  2016-04-02 16:25:54.049   
4               15  2014-11-20 11:34:33.935  2016-04-02 16:25:54.072   
...            ...                      ...                      ...   
105546      150810  2014-11-20 06:16:50.918  2016-04-02 16:31:44.998   
105547      150811  2014-11-20 06:17:39.811   2016-04-02 16:31:44.87   
105548      150812  2014-11-20 06:18:48.894  2016-04-02 16:31:45.009   
105549      150814  2014-11-20 06:21:12.691  2016-04-02 16:31:45.151   
105550      150815  2014-11-20 06:23:12.728  2016-04-02 16:31:46.012   

            clickability_test_id                           excerpt  \
0       546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1       546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2       546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3       546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4       546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                          ...                               ...   
105546  546d86879ad54ec65b000038  Things that matter. Pass 'em on.   
105547  546d86879ad54ec65b000038  Things that matter. Pass 'em on.   
105548  546d86879ad54ec65b000038  Things that matter. Pass 'em on.   
105549  546d86879ad54ec65b000038  Things that matter. Pass 'em on.   
105550  546d86879ad54ec65b000038  Things that matter. Pass 'em on.   

                                                 headline  \
0       Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1       People Sent This Lesbian Questions And Her Rai...   
2       $3 Million Is What It Takes For A State To Leg...   
3       The Fact That Sometimes Innocent People Are Ex...   
4       Reason #351 To End The Death Penalty: It Costs...   
...                                                   ...   
105546  Life For The People Of North Korea Is Worse Th...   
105547  Kim Jong Un Would Really Hate For You To Watch...   
105548  When She Was 9 Years Old, This Girl Saw A Woma...   
105549  This Incredible Young Woman Walked Through The...   
105550  She Took The Stage To Ask The World To Shed Li...   

                                                     lede  \
0       <p>Iff you start with the basic fact that inno...   
1           <p>I'll be honest. I've wondered about 7.</p>   
2       <p>Iff you start with the basic fact that inno...   
3       <p>Iff you start with the basic fact that inno...   
4       <p>Iff you start with the basic fact that inno...   
...                                                   ...   
105546  <p>At age 21, Yeonmi Park is a brand new activ...   
105547  <p>At age 21, Yeonmi Park is a brand new activ...   
105548  <p>At age 21, Yeonmi Park is a brand new activ...   
105549  <p>At age 21, Yeonmi Park is a brand new activ...   
105550  <p>At age 21, Yeonmi Park is a brand new activ...   

                                                     slug  \
0       let-s-see-hire-cops-pay-teachers-buy-books-for...   
1       people-sent-this-lesbian-questions-and-her-rai...   
2       3-million-is-what-it-takes-for-a-state-to-lega...   
3       the-fact-that-sometimes-innocent-people-are-ex...   
4       reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                   ...   
105546  life-for-the-people-of-north-korea-is-worse-th...   
105547  kim-jong-un-would-really-hate-for-you-to-watch...   
105548  when-she-was-9-years-old-this-girl-saw-a-woman...   
105549  this-incredible-young-woman-walked-through-the...   
105550  she-took-the-stage-to-ask-the-world-to-shed-li...   

                   eyecatcher_id  impressions  clicks  significance  \
0       546dce659ad54ec65b000041

We can see that the dataset is only a partial dataset. We proceed by combining the datasets to get a complete dataset.

In [5]:
confirmatory = pd.read_csv("/Users/gohjunwei/Dropbox/4 - upworthy (mike)/Upworthy data documentation/upworthy-archive-datasets/upworthy-archive-confirmatory-packages-03.12.2020.csv")
holdout = pd.read_csv("/Users/gohjunwei/Dropbox/4 - upworthy (mike)/Upworthy data documentation/upworthy-archive-datasets/upworthy-archive-holdout-packages-03.12.2020.csv")

/var/folders/1j/m3_4482d21x5c4l8wtby4flr0000gn/T/ipykernel_42406/382507864.py:1: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  confirmatory = pd.read_csv("/Users/gohjunwei/Dropbox/4 - upworthy (mike)/Upworthy data documentation/upworthy-archive-datasets/upworthy-archive-confirmatory-packages-03.12.2020.csv")


We combine the datasets into a single dataset.

In [6]:
df = pd.concat([exploratory, confirmatory, holdout])

We rename the first column to `id` for easier reference. Each ID represents a separate test. Here, we

In [7]:
df.rename(columns={exploratory.columns[0]: "id"}, inplace=True)

In [8]:
df

id               created_at               updated_at  \
0          11  2014-11-20 11:33:26.475  2016-04-02 16:25:54.046   
1          12  2014-11-20 15:00:01.032  2016-04-02 16:25:54.128   
2          13  2014-11-20 11:33:51.973  2016-04-02 16:25:54.069   
3          14  2014-11-20 11:34:12.107  2016-04-02 16:25:54.049   
4          15  2014-11-20 11:34:33.935  2016-04-02 16:25:54.072   
...       ...                      ...                      ...   
22595  150793  2014-11-20 02:52:17.483   2016-04-02 16:31:40.69   
22596  150806  2014-11-20 04:48:29.008  2016-04-02 16:25:54.006   
22597  150807   2014-11-20 04:48:37.59  2016-04-02 16:25:54.009   
22598  150808  2014-11-20 04:48:52.325  2016-04-02 16:25:54.013   
22599  150809  2014-11-20 04:49:19.374  2016-04-02 16:25:54.016   

           clickability_test_id                           excerpt  \
0      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1      546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                         ...                               ...   
22595  546d564284ad3817bc000025  Things that matter. Pass 'em on.   
22596  546d702084ad380752000038  Things that matter. Pass 'em on.   
22597  546d702084ad380752000038  Things that matter. Pass 'em on.   
22598  546d702084ad380752000038  Things that matter. Pass 'em on.   
22599  546d702084ad380752000038  Things that matter. Pass 'em on.   

                                                headline  \
0      Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1      People Sent This Lesbian Questions And Her Rai...   
2      $3 Million Is What It Takes For A State To Leg...   
3      The Fact That Sometimes Innocent People Are Ex...   
4      Reason #351 To End The Death Penalty: It Costs...   
...                                                  ...   
22595  An Artist Animates The Ugly Truth We Should Al...   
22596  They're Being Called 'Walmart's Worst Nightmar...   
22597  They're Being Called 'Walmart's Worst Nightmar...   
22598  They're Being Called 'Walmart's Worst Nightmar...   
22599  They're Being Called 'Walmart's Worst Nightmar...   

                                                    lede  \
0      <p>Iff you start with the basic fact that inno...   
1          <p>I'll be honest. I've wondered about 7.</p>   
2      <p>Iff you start with the basic fact that inno...   
3      <p>Iff you start with the basic fact that inno...   
4      <p>Iff you start with the basic fact that inno...   
...                                                  ...   
22595  <p>Sad but true: It took the deaths of several...   
22596  <p>When I saw *why* people are calling them "W...   
22597  <p>When I saw *why* people are calling them "W...   
22598  <p>When I saw *why* people are calling them "W...   
22599  <p>When I saw *why* people are calling them "W...   

                                                    slug  \
0      let-s-see-hire-cops-pay-teachers-buy-books-for...   
1      people-sent-this-lesbian-questions-and-her-rai...   
2      3-million-is-what-it-takes-for-a-state-to-lega...   
3      the-fact-that-sometimes-innocent-people-are-ex...   
4      reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                  ...   
22595  an-artist-animates-the-ugly-truth-we-should-al...   
22596  theyre-being-called-walmarts-worst-nightmare-a...   
22597  theyre-being-called-walmarts-worst-nightmare-a...   
22598  theyre-being-called-walmarts-worst-nightmare-a...   
22599  theyre-being-called-walmarts-worst-nightmare-a...   

                  eyecatcher_id  impressions  clicks  significance  \
0      546dce659ad54ec65b000041         3118       8           0.1   
1      546d1b4bfd3617f091000041         4587     130          55.8   
2   

Here, we see that there are a total of 311934 unique tests in the dataset.

In [9]:
exploratory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105551 entries, 0 to 105550
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Unnamed: 0            105551 non-null  int64  
 1   created_at            105551 non-null  object 
 2   updated_at            105551 non-null  object 
 3   clickability_test_id  105551 non-null  object 
 4   excerpt               94335 non-null   object 
 5   headline              105551 non-null  object 
 6   lede                  105485 non-null  object 
 7   slug                  105551 non-null  object 
 8   eyecatcher_id         105420 non-null  object 
 9   impressions           105551 non-null  int64  
 10  clicks                105551 non-null  int64  
 11  significance          105551 non-null  float64
 12  first_place           105551 non-null  bool   
 13  winner                105551 non-null  bool   
 14  share_text            14632 non-null   object 
 15  

- **`created_at`**: Time the package was created (timezone unknown). This column records when each content package was initially created in the system.

- **`test_week`**: Week the package was created, a variable constructed by the archive creators for stratified random sampling. This helps in organizing and analyzing the data by specific time periods.

- **`clickability_test_id`**: The test ID. Viewers were randomly assigned to packages with the same test ID, allowing for controlled experiments on clickability.

- **`impressions`**: The number of viewers who were assigned to this package. The total number of participants for a given test is the sum of impressions for all packages that share the same clickability_test_id.

- **`headline`**: The headline being tested. This is a critical component of the content package as it's a major factor influencing click-through rates.

- **`eyecatcher_id`**: Image ID. Image files are not available. Packages that shared the same image have the same eyecatcher_id, indicating they used the same visual element.

- **`clicks`**: The number of viewers (impressions) that clicked on the package. The click rate for a given package is the number of clicks divided by the number of impressions.

- **`excerpt`**: Article excerpt. This may provide additional context or teaser information about the content package.

- **`lede`**: The opening sentence or paragraph of the story, often used to hook the reader.

- **`slug`**: Internal name for the web address. This is typically a URL-friendly version of the headline or a key aspect of the content.

- **`share_text`**: Summary for display on social media when the article is shared. This was not shown in tests since tests were conducted on the Upworthy website.

- **`square`**: When used, part of the same social media sharing suggestion as the share text. It might be a specific format or template for sharing content.

- **`significance`**: NOT an estimate of statistical significance; a complex, inconsistent calculation that compared the clicks on a package to the clicks on all previous packages that were fielded on the same pages.

- **`first_place`**: Along with significance, shown to editors to guide decisions about what test to choose. It indicates whether a package was initially considered the most promising based on certain metrics.

- **`winner`**: Whether a package was selected by editors to be used on the Upworthy site after the test. This reflects the outcome of the testing process in terms of content selection.

- **`updated_at`**: The last time the package was updated in the Upworthy system. This tracks any changes or updates made to the package after its initial creation.

We confirm that all tests have a unique ID.

In [10]:
df["id"].nunique()

128151

We remove all duplicate tests from the dataset.

In [11]:
df.drop_duplicates(subset="id", inplace=True)

In [12]:
df

id               created_at               updated_at  \
0          11  2014-11-20 11:33:26.475  2016-04-02 16:25:54.046   
1          12  2014-11-20 15:00:01.032  2016-04-02 16:25:54.128   
2          13  2014-11-20 11:33:51.973  2016-04-02 16:25:54.069   
3          14  2014-11-20 11:34:12.107  2016-04-02 16:25:54.049   
4          15  2014-11-20 11:34:33.935  2016-04-02 16:25:54.072   
...       ...                      ...                      ...   
22595  150793  2014-11-20 02:52:17.483   2016-04-02 16:31:40.69   
22596  150806  2014-11-20 04:48:29.008  2016-04-02 16:25:54.006   
22597  150807   2014-11-20 04:48:37.59  2016-04-02 16:25:54.009   
22598  150808  2014-11-20 04:48:52.325  2016-04-02 16:25:54.013   
22599  150809  2014-11-20 04:49:19.374  2016-04-02 16:25:54.016   

           clickability_test_id                           excerpt  \
0      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1      546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                         ...                               ...   
22595  546d564284ad3817bc000025  Things that matter. Pass 'em on.   
22596  546d702084ad380752000038  Things that matter. Pass 'em on.   
22597  546d702084ad380752000038  Things that matter. Pass 'em on.   
22598  546d702084ad380752000038  Things that matter. Pass 'em on.   
22599  546d702084ad380752000038  Things that matter. Pass 'em on.   

                                                headline  \
0      Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1      People Sent This Lesbian Questions And Her Rai...   
2      $3 Million Is What It Takes For A State To Leg...   
3      The Fact That Sometimes Innocent People Are Ex...   
4      Reason #351 To End The Death Penalty: It Costs...   
...                                                  ...   
22595  An Artist Animates The Ugly Truth We Should Al...   
22596  They're Being Called 'Walmart's Worst Nightmar...   
22597  They're Being Called 'Walmart's Worst Nightmar...   
22598  They're Being Called 'Walmart's Worst Nightmar...   
22599  They're Being Called 'Walmart's Worst Nightmar...   

                                                    lede  \
0      <p>Iff you start with the basic fact that inno...   
1          <p>I'll be honest. I've wondered about 7.</p>   
2      <p>Iff you start with the basic fact that inno...   
3      <p>Iff you start with the basic fact that inno...   
4      <p>Iff you start with the basic fact that inno...   
...                                                  ...   
22595  <p>Sad but true: It took the deaths of several...   
22596  <p>When I saw *why* people are calling them "W...   
22597  <p>When I saw *why* people are calling them "W...   
22598  <p>When I saw *why* people are calling them "W...   
22599  <p>When I saw *why* people are calling them "W...   

                                                    slug  \
0      let-s-see-hire-cops-pay-teachers-buy-books-for...   
1      people-sent-this-lesbian-questions-and-her-rai...   
2      3-million-is-what-it-takes-for-a-state-to-lega...   
3      the-fact-that-sometimes-innocent-people-are-ex...   
4      reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                  ...   
22595  an-artist-animates-the-ugly-truth-we-should-al...   
22596  theyre-being-called-walmarts-worst-nightmare-a...   
22597  theyre-being-called-walmarts-worst-nightmare-a...   
22598  theyre-being-called-walmarts-worst-nightmare-a...   
22599  theyre-being-called-walmarts-worst-nightmare-a...   

                  eyecatcher_id  impressions  clicks  significance  \
0      546dce659ad54ec65b000041         3118       8           0.1   
1      546d1b4bfd3617f091000041         4587     130          55.8   
2   

We now determine the total number of unique tests in the dataset.

In [13]:
df["clickability_test_id"].nunique()

27614

Now, we remove all tests where there is only one experiment. In other words, if count of `clickability_test_id` is 1, we remove the test.

In [14]:
df = df.groupby("clickability_test_id").filter(lambda x: len(x) > 1)

In [15]:
df

id               created_at               updated_at  \
0          11  2014-11-20 11:33:26.475  2016-04-02 16:25:54.046   
1          12  2014-11-20 15:00:01.032  2016-04-02 16:25:54.128   
2          13  2014-11-20 11:33:51.973  2016-04-02 16:25:54.069   
3          14  2014-11-20 11:34:12.107  2016-04-02 16:25:54.049   
4          15  2014-11-20 11:34:33.935  2016-04-02 16:25:54.072   
...       ...                      ...                      ...   
22595  150793  2014-11-20 02:52:17.483   2016-04-02 16:31:40.69   
22596  150806  2014-11-20 04:48:29.008  2016-04-02 16:25:54.006   
22597  150807   2014-11-20 04:48:37.59  2016-04-02 16:25:54.009   
22598  150808  2014-11-20 04:48:52.325  2016-04-02 16:25:54.013   
22599  150809  2014-11-20 04:49:19.374  2016-04-02 16:25:54.016   

           clickability_test_id                           excerpt  \
0      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1      546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                         ...                               ...   
22595  546d564284ad3817bc000025  Things that matter. Pass 'em on.   
22596  546d702084ad380752000038  Things that matter. Pass 'em on.   
22597  546d702084ad380752000038  Things that matter. Pass 'em on.   
22598  546d702084ad380752000038  Things that matter. Pass 'em on.   
22599  546d702084ad380752000038  Things that matter. Pass 'em on.   

                                                headline  \
0      Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1      People Sent This Lesbian Questions And Her Rai...   
2      $3 Million Is What It Takes For A State To Leg...   
3      The Fact That Sometimes Innocent People Are Ex...   
4      Reason #351 To End The Death Penalty: It Costs...   
...                                                  ...   
22595  An Artist Animates The Ugly Truth We Should Al...   
22596  They're Being Called 'Walmart's Worst Nightmar...   
22597  They're Being Called 'Walmart's Worst Nightmar...   
22598  They're Being Called 'Walmart's Worst Nightmar...   
22599  They're Being Called 'Walmart's Worst Nightmar...   

                                                    lede  \
0      <p>Iff you start with the basic fact that inno...   
1          <p>I'll be honest. I've wondered about 7.</p>   
2      <p>Iff you start with the basic fact that inno...   
3      <p>Iff you start with the basic fact that inno...   
4      <p>Iff you start with the basic fact that inno...   
...                                                  ...   
22595  <p>Sad but true: It took the deaths of several...   
22596  <p>When I saw *why* people are calling them "W...   
22597  <p>When I saw *why* people are calling them "W...   
22598  <p>When I saw *why* people are calling them "W...   
22599  <p>When I saw *why* people are calling them "W...   

                                                    slug  \
0      let-s-see-hire-cops-pay-teachers-buy-books-for...   
1      people-sent-this-lesbian-questions-and-her-rai...   
2      3-million-is-what-it-takes-for-a-state-to-lega...   
3      the-fact-that-sometimes-innocent-people-are-ex...   
4      reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                  ...   
22595  an-artist-animates-the-ugly-truth-we-should-al...   
22596  theyre-being-called-walmarts-worst-nightmare-a...   
22597  theyre-being-called-walmarts-worst-nightmare-a...   
22598  theyre-being-called-walmarts-worst-nightmare-a...   
22599  theyre-being-called-walmarts-worst-nightmare-a...   

                  eyecatcher_id  impressions  clicks  significance  \
0      546dce659ad54ec65b000041         3118       8           0.1   
1      546d1b4bfd3617f091000041         4587     130          55.8   
2   

We begin by cleaning the data. First, we calculate the CTR of each content piece by dividing the number of clicks by the number of impressions.

In [16]:
df["clickability_test_id"].nunique()

27612

We end up with 27612 unique tests that we can draw from.

In [17]:
df["CTR"] = df["clicks"] / df["impressions"]

In [18]:
df

id               created_at               updated_at  \
0          11  2014-11-20 11:33:26.475  2016-04-02 16:25:54.046   
1          12  2014-11-20 15:00:01.032  2016-04-02 16:25:54.128   
2          13  2014-11-20 11:33:51.973  2016-04-02 16:25:54.069   
3          14  2014-11-20 11:34:12.107  2016-04-02 16:25:54.049   
4          15  2014-11-20 11:34:33.935  2016-04-02 16:25:54.072   
...       ...                      ...                      ...   
22595  150793  2014-11-20 02:52:17.483   2016-04-02 16:31:40.69   
22596  150806  2014-11-20 04:48:29.008  2016-04-02 16:25:54.006   
22597  150807   2014-11-20 04:48:37.59  2016-04-02 16:25:54.009   
22598  150808  2014-11-20 04:48:52.325  2016-04-02 16:25:54.013   
22599  150809  2014-11-20 04:49:19.374  2016-04-02 16:25:54.016   

           clickability_test_id                           excerpt  \
0      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1      546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                         ...                               ...   
22595  546d564284ad3817bc000025  Things that matter. Pass 'em on.   
22596  546d702084ad380752000038  Things that matter. Pass 'em on.   
22597  546d702084ad380752000038  Things that matter. Pass 'em on.   
22598  546d702084ad380752000038  Things that matter. Pass 'em on.   
22599  546d702084ad380752000038  Things that matter. Pass 'em on.   

                                                headline  \
0      Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1      People Sent This Lesbian Questions And Her Rai...   
2      $3 Million Is What It Takes For A State To Leg...   
3      The Fact That Sometimes Innocent People Are Ex...   
4      Reason #351 To End The Death Penalty: It Costs...   
...                                                  ...   
22595  An Artist Animates The Ugly Truth We Should Al...   
22596  They're Being Called 'Walmart's Worst Nightmar...   
22597  They're Being Called 'Walmart's Worst Nightmar...   
22598  They're Being Called 'Walmart's Worst Nightmar...   
22599  They're Being Called 'Walmart's Worst Nightmar...   

                                                    lede  \
0      <p>Iff you start with the basic fact that inno...   
1          <p>I'll be honest. I've wondered about 7.</p>   
2      <p>Iff you start with the basic fact that inno...   
3      <p>Iff you start with the basic fact that inno...   
4      <p>Iff you start with the basic fact that inno...   
...                                                  ...   
22595  <p>Sad but true: It took the deaths of several...   
22596  <p>When I saw *why* people are calling them "W...   
22597  <p>When I saw *why* people are calling them "W...   
22598  <p>When I saw *why* people are calling them "W...   
22599  <p>When I saw *why* people are calling them "W...   

                                                    slug  \
0      let-s-see-hire-cops-pay-teachers-buy-books-for...   
1      people-sent-this-lesbian-questions-and-her-rai...   
2      3-million-is-what-it-takes-for-a-state-to-lega...   
3      the-fact-that-sometimes-innocent-people-are-ex...   
4      reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                  ...   
22595  an-artist-animates-the-ugly-truth-we-should-al...   
22596  theyre-being-called-walmarts-worst-nightmare-a...   
22597  theyre-being-called-walmarts-worst-nightmare-a...   
22598  theyre-being-called-walmarts-worst-nightmare-a...   
22599  theyre-being-called-walmarts-worst-nightmare-a...   

                  eyecatcher_id  impressions  clicks  significance  \
0      546dce659ad54ec65b000041         3118       8           0.1   
1      546d1b4bfd3617f091000041         4587     130          55.8   
2   

We now proceed to remove all tests where there are variations in the image. We do this by removing all tests where there are variations in the `eyecatcher_id`.

In [19]:
df = df.groupby("clickability_test_id").filter(lambda x: len(x["eyecatcher_id"].unique()) == 1)

In [20]:
df

id               created_at               updated_at  \
0          11  2014-11-20 11:33:26.475  2016-04-02 16:25:54.046   
1          12  2014-11-20 15:00:01.032  2016-04-02 16:25:54.128   
2          13  2014-11-20 11:33:51.973  2016-04-02 16:25:54.069   
3          14  2014-11-20 11:34:12.107  2016-04-02 16:25:54.049   
4          15  2014-11-20 11:34:33.935  2016-04-02 16:25:54.072   
...       ...                      ...                      ...   
22587  150778  2014-11-20 02:12:35.923  2016-04-02 16:25:53.927   
22588  150779  2014-11-20 02:12:35.963  2016-04-02 16:31:40.659   
22589  150780  2014-11-20 02:12:36.023   2016-04-02 16:25:53.96   
22590  150781  2014-11-20 02:12:36.114  2016-04-02 16:25:53.929   
22591  150782  2014-11-20 02:12:36.241  2016-04-02 16:25:53.932   

           clickability_test_id                           excerpt  \
0      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1      546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                         ...                               ...   
22587  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22588  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22589  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22590  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22591  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   

                                                headline  \
0      Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1      People Sent This Lesbian Questions And Her Rai...   
2      $3 Million Is What It Takes For A State To Leg...   
3      The Fact That Sometimes Innocent People Are Ex...   
4      Reason #351 To End The Death Penalty: It Costs...   
...                                                  ...   
22587  Say Goodbye To The News You Once Knew. These 2...   
22588  Say Goodbye To The News You Once Knew. These 2...   
22589  Say Goodbye To The News You Once Knew. These 2...   
22590  Say Goodbye To The News You Once Knew. These 2...   
22591  Say Goodbye To The News You Once Knew. These 2...   

                                                    lede  \
0      <p>Iff you start with the basic fact that inno...   
1          <p>I'll be honest. I've wondered about 7.</p>   
2      <p>Iff you start with the basic fact that inno...   
3      <p>Iff you start with the basic fact that inno...   
4      <p>Iff you start with the basic fact that inno...   
...                                                  ...   
22587  <p>This is a pretty great idea for a TV show! ...   
22588  <p>This is a pretty great idea for a TV show! ...   
22589  <p>This is a pretty great idea for a TV show! ...   
22590  <p>This is a pretty great idea for a TV show! ...   
22591  <p>This is a pretty great idea for a TV show! ...   

                                                    slug  \
0      let-s-see-hire-cops-pay-teachers-buy-books-for...   
1      people-sent-this-lesbian-questions-and-her-rai...   
2      3-million-is-what-it-takes-for-a-state-to-lega...   
3      the-fact-that-sometimes-innocent-people-are-ex...   
4      reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                  ...   
22587  say-goodbye-to-the-news-you-once-knew-these-2-...   
22588  say-goodbye-to-the-news-you-once-knew-these-2-...   
22589  say-goodbye-to-the-news-you-once-knew-these-2-...   
22590  say-goodbye-to-the-news-you-once-knew-these-2-...   
22591  say-goodbye-to-the-news-you-once-knew-these-2-...   

                  eyecatcher_id  impressions  clicks  significance  \
0      546dce659ad54ec65b000041         3118       8           0.1   
1      546d1b4bfd3617f091000041         4587     130          55.8   
2   

In [21]:
df["clickability_test_id"].nunique()

17882

We end up with 17882 unique tests that we can draw from.

In [22]:
df['headline'] = df['headline'].astype(str)
df['lede'] = df['lede'].astype(str)
df['slug'] = df['slug'].astype(str)

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 81816 entries, 0 to 22591
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    81816 non-null  int64  
 1   created_at            81816 non-null  object 
 2   updated_at            81816 non-null  object 
 3   clickability_test_id  81816 non-null  object 
 4   excerpt               73130 non-null  object 
 5   headline              81816 non-null  object 
 6   lede                  81816 non-null  object 
 7   slug                  81816 non-null  object 
 8   eyecatcher_id         81660 non-null  object 
 9   impressions           81816 non-null  int64  
 10  clicks                81816 non-null  int64  
 11  significance          81816 non-null  float64
 12  first_place           81816 non-null  bool   
 13  winner                81816 non-null  bool   
 14  share_text            13721 non-null  object 
 15  square                26

We now have a final dataset that we can use. This will be the raw source of truth for all future analysis.

In [24]:
df['created_at'] = pd.to_datetime(df['created_at'], errors = 'coerce')
df['updated_at'] = pd.to_datetime(df['updated_at'], errors = 'coerce')

In [25]:
df['timedelta'] = df['updated_at'] - df['created_at']

In [26]:
df['timedelta_numeric'] = pd.to_timedelta(df['timedelta']).dt.total_seconds()

In [27]:
from sklearn.preprocessing import MinMaxScaler

timedelta_array = df[['timedelta_numeric']].values  # Convert to a 2D array
scaler = MinMaxScaler()
df['timedelta_minmax'] = scaler.fit_transform(timedelta_array)

In [28]:
df.to_csv('df_cleaned.csv', index=False)

In [29]:
df = df.dropna(subset = ['timedelta_minmax'])

In [30]:
df

id              created_at              updated_at  \
0          11 2014-11-20 11:33:26.475 2016-04-02 16:25:54.046   
1          12 2014-11-20 15:00:01.032 2016-04-02 16:25:54.128   
2          13 2014-11-20 11:33:51.973 2016-04-02 16:25:54.069   
3          14 2014-11-20 11:34:12.107 2016-04-02 16:25:54.049   
4          15 2014-11-20 11:34:33.935 2016-04-02 16:25:54.072   
...       ...                     ...                     ...   
22587  150778 2014-11-20 02:12:35.923 2016-04-02 16:25:53.927   
22588  150779 2014-11-20 02:12:35.963 2016-04-02 16:31:40.659   
22589  150780 2014-11-20 02:12:36.023 2016-04-02 16:25:53.960   
22590  150781 2014-11-20 02:12:36.114 2016-04-02 16:25:53.929   
22591  150782 2014-11-20 02:12:36.241 2016-04-02 16:25:53.932   

           clickability_test_id                           excerpt  \
0      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
1      546e01d626714c6c4400004e  Things that matter. Pass 'em on.   
2      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
3      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
4      546dd17e26714c82cc00001c  Things that matter. Pass 'em on.   
...                         ...                               ...   
22587  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22588  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22589  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22590  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   
22591  546d4e139ad54eec8d000019  Things that matter. Pass 'em on.   

                                                headline  \
0      Let’s See … Hire Cops, Pay Teachers, Buy Books...   
1      People Sent This Lesbian Questions And Her Rai...   
2      $3 Million Is What It Takes For A State To Leg...   
3      The Fact That Sometimes Innocent People Are Ex...   
4      Reason #351 To End The Death Penalty: It Costs...   
...                                                  ...   
22587  Say Goodbye To The News You Once Knew. These 2...   
22588  Say Goodbye To The News You Once Knew. These 2...   
22589  Say Goodbye To The News You Once Knew. These 2...   
22590  Say Goodbye To The News You Once Knew. These 2...   
22591  Say Goodbye To The News You Once Knew. These 2...   

                                                    lede  \
0      <p>Iff you start with the basic fact that inno...   
1          <p>I'll be honest. I've wondered about 7.</p>   
2      <p>Iff you start with the basic fact that inno...   
3      <p>Iff you start with the basic fact that inno...   
4      <p>Iff you start with the basic fact that inno...   
...                                                  ...   
22587  <p>This is a pretty great idea for a TV show! ...   
22588  <p>This is a pretty great idea for a TV show! ...   
22589  <p>This is a pretty great idea for a TV show! ...   
22590  <p>This is a pretty great idea for a TV show! ...   
22591  <p>This is a pretty great idea for a TV show! ...   

                                                    slug  \
0      let-s-see-hire-cops-pay-teachers-buy-books-for...   
1      people-sent-this-lesbian-questions-and-her-rai...   
2      3-million-is-what-it-takes-for-a-state-to-lega...   
3      the-fact-that-sometimes-innocent-people-are-ex...   
4      reason-351-to-end-the-death-penalty-it-costs-3...   
...                                                  ...   
22587  say-goodbye-to-the-news-you-once-knew-these-2-...   
22588  say-goodbye-to-the-news-you-once-knew-these-2-...   
22589  say-goodbye-to-the-news-you-once-knew-these-2-...   
22590  say-goodbye-to-the-news-you-once-knew-these-2-...   
22591  say-goodbye-to-the-news-you-once-knew-these-2-...   

                  eyecatcher_id  impressions  ...  significance  first_place  \
0      546dce659ad54ec65b000041         3118  ...           0.1        False   
1      546d1b4bfd3617f091000041         4587  ...          55.8        False  

We now calculate clicks_rl, such that:

clicks_rl ~ Binomial(impressions_rl, θ_rl)

In [31]:
from sklearn.model_selection import train_test_split
X = df.drop(['CTR'], axis = 1)
y = df['CTR']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.to_csv('df_cleaned_X_train.csv', index=False)
X_test.to_csv('df_cleaned_X_test.csv', index=False)

y_train.to_csv('df_cleaned_y_train.csv', index=False)
y_test.to_csv('df_cleaned_y_test.csv', index=False)

In [32]:
len(X_train)

62804

In [33]:
len(X_test)

15702